Added Epson projector support

pszafer opened the pull request:
## Description:

Hello,
I created epson projector support.
It is based on Android app (packet sniffing) from epson.
Tested with Epson EH-TW5350


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5005

I can make documentation if you will to add this module to home assistant.

## Example entry for `configuration.yaml` (if applicable):
```yaml
- platform: epson
  host: 192.168.1.2
  name: Epson Projector
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

pszafer committed:
Epson projector added

pszafer committed:
Epson projector added

pszafer committed:
added support to get CMODE mode

pszafer committed:
initial release of epson module

pszafer committed:
epson

homeassistant commented:
Hi @pszafer, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#13416).

Once you do that we will be able to review and accept this pull request.

Thanks!

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+"""
+Support for Epson projector
```
Add a period at the end please.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+Support for Epson projector
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/
```
Please add the link to the documentation.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/
+
```
Remove the blank line

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_STEP, SUPPORT_NEXT_TRACK, SUPPORT_PREVIOUS_TRACK
+)
+
+""" ,SUPPORT_PLAY, SUPPORT_PAUSE, SUPPORT_STOP"""
```
Remove this.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+from homeassistant.helpers.aiohttp_client import async_create_clientsession
+from homeassistant.const import (ATTR_ENTITY_ID, CONF_HOST,CONF_NAME,CONF_PORT,CONF_SSL, STATE_ON, STATE_OFF, STATE_UNKNOWN, STATE_PLAYING)
+
+import homeassistant.util as util
```
Keep the import [ordered](https://www.python.org/dev/peps/pep-0008/#imports). `isort` can help.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+
+import homeassistant.util as util
+
+DEPENDENCIES = []
```
Can be removed.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME) : cv.string,
+    vol.Optional(CONF_PORT, default=80) : cv.port,
+    vol.Optional(CONF_SSL, default=False) : cv.boolean
+})
```
No whitespace before `:`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+
```
Docstring is missing.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+    epson.update()
+    hass.data[DATA_EPSON].append(epson)
+    async_add_devices([epson], update_before_add=True)
+    # async_add_devices([epson])
```
Remove this.

pszafer replied:
do you mean all 4 lines or just commented out line?

I need hass.data[DATA_EPSON] in async_service_handler.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+class EpsonProjector(MediaPlayerDevice):
+    """Representation of Epson Projector Device."""
+
+    def __init__(self, hass, name, host, port, encryption, key_commands):
```
Docstring is missing.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+        self._source_list = list(DEFAULT_SOURCES.values())
+        self.KEY_COMMANDS = key_commands
+        self._encryption = encryption
+        self._state = STATE_UNKNOWN
```
Set it to `None`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+            "Accept-Encoding" : "gzip, deflate",
+            "Accept" : "application/json, text/javascript",
+            "Referer": "{http_protocol}://{host}:{port}/cgi-bin/webconf".format(http_protocol=http_protocol,host=self._host,port=self._port)
+        };
```
Use constants for `Accept`, `application/json`, etc.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+
+        self.websession = async_create_clientsession(hass, verify_ssl=self._encryption)
+        self.websession_action = async_create_clientsession(hass, verify_ssl=self._encryption)
+        self.update()
```
Remove this.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+        return self._source_list
+
+    @property
+    def cmode(self):
```
Docstring is missing.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+        return self._cmode
+
+    @asyncio.coroutine
+    def select_cmode(self, cmode):
```
Dito.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+    @asyncio.coroutine
+    def async_select_source(self, source):
+        """Select input source."""
+        _LOGGER.info("select source")
```
Set this to `_LOGGER.debug()`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+
+    @asyncio.coroutine
+    def async_mute_volume(self, mute):
+        """Mute (true) or unmute (false) projector. It also mutes video input!  """
```
No whitespace around docstrings.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+
+    @asyncio.coroutine
+    def async_media_pause(self):
+        """Play."""
```
Stale.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+        return self.sendCommand("BACK")
+
+    @asyncio.coroutine
+    def sendCommand(self, command):
```
Docstring is missing.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+
+    @asyncio.coroutine
+    def sendCommand(self, command):
+        _LOGGER.info("COMMAND %s", command)
```
Set this to `_LOGGER.debug()`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+        return attributes
+
+
+#curl 'http://192.168.4.131/cgi-bin/json_query?jsoncallback=HDMILINK?%2001&_=1520179300356' -H 'Accept-Encoding: gzip, deflate'  -H 'Accept: application/json, text/javascript, */*; q=0.01' -H 'Referer: http://192.168.4.131/cgi-bin/webconf' -H 'X-Requested-With: XMLHttpRequest' -H 'Connection: keep-alive' --compressed
```
Remove this.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,320 @@
+#     text = config[DOMAIN].get(CONF_TEXT, DEFAULT_TEXT)
+#     hass.states.set('first.Hello_State', text)
+#
+#     return True
```
Stale content.

pszafer committed:
fixes according to @fabaff review.

pszafer committed:
remove STATE_PLAYING and homeassistant.helpers.aiohttp_client.async_create_clientsession

pszafer committed:
docstring fixed

pszafer committed:
snake_case naming style added
repaired missing of async_create_clientsession

pszafer committed:
isorted

pszafer committed:
snake_case second try.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+ATTR_CMODE = 'cmode'
+SUPPORT_CMODE = 33001
+ACCEPT_ENCODING = "gzip, deflate"
+ACCEPT_HEADER = "application/json, text/javascript"
```
Please use the constants from `from aiohttp import hdrs` or `const.py` if possible.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+        hass, name, host,
+        config.get(CONF_PORT), config.get(CONF_SSL), KEY_COMMANDS)
+
+    epson.update()
```
I guess that there will be an exception if the projector is not on/accessible/etc. You need to handle it, e.g., use `PlatformNotReady` or make the setup fail.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle for services."""
+        entity_ids = service.data.get('entity_id')
```
`entity_id` is available in `const.py`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+        self._source_list = list(DEFAULT_SOURCES.values())
+        self.key_commands = key_commands
+        self._encryption = encryption
+        self._state = STATE_UNKNOWN
```
Set it to `None` and let Home Assistant handle it.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+        self._headers = {
+            "Accept-Encoding": ACCEPT_ENCODING,
+            "Accept": ACCEPT_HEADER,
+            "Referer": referer
```
Please use the constants from `from aiohttp import hdrs` or `const.py` if possible.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+                return None
+            return True
+        except (asyncio.TimeoutError, aiohttp.ClientError):
+            _LOGGER.error("[%s] Error getting info", DOMAIN)
```
There is no need to include `DOMAIN`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+        params = self.key_commands[command]
+        try:
+            url = '{url}{type}'.format(url=self._http_url, type='directsend')
+            response = yield from self.websession_action.get(
```
Add a timeout.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+                self._state = STATE_ON
+                self._cmode = CMODE_LIST[reply_code]
+        except aiohttp.ClientError:
+            _LOGGER.error("[%s] Error getting info", DOMAIN)
```
There is no need to include `DOMAIN`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+                    headers=self._headers)
+            if response.status != 200:
+                _LOGGER.warning(
+                    "[%s] Error %d on Epson.", DOMAIN, response.status)
```
There is no need to include `DOMAIN`.

fabaff commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,316 @@
+            verify_ssl=self._encryption)
+        self.websession_action = async_create_clientsession(
+            hass,
+            verify_ssl=self._encryption)
```
Any reason not to share one session?

pszafer replied:
if I share one session and turn on the projector I get error on update status.
Am I doing something wrong with sharing one session?

```
2018-03-27 11:51:31 WARNING (MainThread) [homeassistant.helpers.entity] Update of media_player.epson_projector is taking over 10 seconds
2018-03-27 11:51:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event system_log_event[L]: timestamp=1522144291.5457628, level=WARNING, message=Update of media_player.epson_projector is taking over 10 seconds, exception=, source=core.py>
2018-03-27 11:51:31 ERROR (MainThread) [custom_components.media_player.epson] Error getting info
```

If I have two sessions it looks like this. No error updating info from projector.

```
2018-03-27 11:56:11 WARNING (MainThread) [homeassistant.helpers.entity] Update of media_player.epson_projector is taking over 10 seconds
2018-03-27 11:56:11 INFO (MainThread) [homeassistant.core] Bus:Handling <Event system_log_event[L]: timestamp=1522144571.9009275, level=WARNING, message=Update of media_player.epson_projector is taking over 10 seconds, exception=, source=core.py>
2018-03-27 11:56:16 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_executed[L]: service_call_id=139837689363424-3>
2018-03-27 11:56:59 INFO (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=media_player.epson_projector, old_state=<state media_player.epson_projector=off; cmode=Cinema, friendly_name=Epson Projector, supported_features=36345 @ 2018-03-27T11:55:42.101902+02:00>, new_state=<state media_player.epson_projector=on; source_list=['HDMI1', 'HDMI2', 'PC', 'VIDEO', 'USB', 'LAN', 'WiFi Direct'], cmode=Cinema, friendly_name=Epson Projector, supported_features=36345 @ 2018-03-27T11:56:59.057799+02:00>>
```

pszafer committed:
Included @fabaff review from 26/03/2018

pszafer committed:
deleted timedelta

dthulke commented:
I tested your component with my EH-TW6700 (I also started creating a platform but didn't yet found the time to finish it). The only issue I found in your component was that the component crashes if an unknown cmode is returned by the beamer (in my case `00: Auto`). Everything else works as expected.

The commands the different projector models accept, including all available cmodes, are documented in the following table: ftp://download.epson-europe.com/pub/download/3756/epson375633eu.xlsx

Maybe you can just add all cmodes listed in the document.
There are also commands to query the current lamp hours, the current input and current volume.

pszafer commented:
@dthulke ok, I will try to do that. Cool documentation btw... I asked Epson support about month ago about documentation of projector functions and they said they don't have one... I made it all by sniffing and this doc seems to be much easier way to do all of that.

I'll try to rewrite my code to follow this documentation

pszafer committed:
switched to mute audio only
added support for more CMODE's
checking current selected source

pszafer committed:
switched to single websession.
It is generating error during update.

pszafer committed:
added extra CMODEs

pszafer committed:
getProperty to get_property

pszafer committed:
checking if module know source type and cmode type

pszafer commented:
@dthulke I added all cmode I could find in documentation xls.

> There are also commands to query the current lamp hours, the current input and current volume.
I added source state.

Current volume returning always '121' value for me. Even if I play video and sound through projector.
I can retrieve this status, but I don't know if this make sense.
Can you check if it is giving you any reasonable answer

```
curl 'http://PROJECTORIP/cgi-bin/json_query?jsoncallback=VOL?' -H 'Accept-Encoding: gzip, deflate' -H 'Accept: application/json, text/javascript' -H 'Referer: http://PROJECTORIP/cgi-bin/webconf
```

dthulke commented:
I can't reproduce the issue. My device always returns values between 0 and 255 (as documented) even with different inputs or no input at all.

What happens if you try to set the volume via the VOL command? For example as follows:
`curl -v --header 'Referer:http://PROJECTORIP/cgi-bin/webconf' 'http://PROJECTORIP/cgi-bin/directsend?VOL=0'`

pszafer commented:
I will add volume status checking.
It doesn't really matter for me.
Conversation looks like this. 
```                                                                                                               
pszafer@TBH105:~$ curl 'http://192.168.4.131/cgi-bin/json_query?jsoncallback=VOL?' -H 'Accept-Encoding: gzip, deflate' -H 'Referer: http://192.168.4.131/cgi-bin/webcon
f'                                                                                                                     
{                                                                                                                                                                                                                   
"projector": {                                                                                                                                                                                                      
"feature": {                                                                                                                                                                                                        
"name": "esc/vp21",                                                                
"query": "VOL?",                                                                                                                                                                                                    
"reply": "121",                                                                                                                                                                                                     
"error": false                                                                                                                                                                                                      
}                                                                                                                                                                                                                   
}                                                                                                                                                                                                                   
}    
pszafer@TBH105:~$ curl 'http://192.168.4.131/cgi-bin/directsend?VOL=0' -H 'Accept-Encoding: gzip, deflate' -H 'Referer: http://192.168.4.131/cgi-bin/webconf'                                                                                                                                                                                                                   
pszafer@TBH105:~$ curl 'http://192.168.4.131/cgi-bin/json_query?jsoncallback=VOL?' -H 'Accept-Encoding: gzip, deflate' -H 'Accept: application/json, text/javascript, */*; q=0.01' -H 'Referer: http://192.168.4.131/cgi-bin/webconf'                                                                                                   
{                                                                                                                                                                                                                   
"projector": {                                                                                                                                                                                                      
"feature": {                                                                                                                                                                                                        
"name": "esc/vp21",        
"query": "VOL?",                                                                                                                                                                                                    
"reply": "121",                                                                                                                                                                                                     
"error": false             
}                                                                                                                                                                                                                   
}                                                                                                                                                                                                                   
}                            
```

pszafer commented:
Hi @fabaff , will you merge this pull request or is anything else I need to do with that module?

balloob commented on homeassistant/components/media_player/epson.py:
```diff
@@ -0,0 +1,419 @@
+            self._state = STATE_OFF
+
+    @asyncio.coroutine
+    def get_property(self, command):
```
This is too much protocol specific information. Please extract this into a third party library and have the platform use that. This should include all key mappings and cmodes etc

balloob reviewed (changes requested):
Please extract interaction logic into a third party lib.

pszafer commented:
Done. Logic moved to external package.

dgomes commented:
You will need to rebase...

pszafer commented:
What am I doing wrong trying to upgrade my repo that git is showing so many changes?

I did:

```
git fetch upstream
git rebase upstream/dev
```

and for homeassistant/const.py
git mergetool and bumped it to version 0.71.0.dev0

balloob commented:
It seems like you have some commits in this branch that are not yours, please clean it up.

pszafer committed:
moved epson logic to external modul

pszafer committed:
moved import 3rd party library to method

pszafer committed:
moved imported vars to methods

pszafer committed:
Travis required Please run script/gen_requirements_all.py
